lib/ansible/modules/cloud/docker/docker_swarm.py

```diff
@@ -0,0 +1,524 @@
+try:
+    from docker.errors import APIError
+except ImportError:
+    # missing docker-py handled in ansible.module_utils.docker
+    pass
+
+from ansible.module_utils.docker_common import AnsibleDockerClient, DockerBaseClass
+
+
+class TaskParameters(DockerBaseClass):
+    def __init__(self, client):
+        super(TaskParameters, self).__init__()
+
+        self.state = None
+        self.advertise_addr = None
+        self.listen_addr = None
+        self.force_new_cluster = None
+        self.remote_addrs = None
+        self.join_token = None
+
+        # Spec
+        self.snapshot_interval = None
+        self.task_history_retention_limit = None
+        self.keep_old_snapshots = None
+        self.log_entries_for_slow_followers = None
+        self.heartbeat_tick = None
+        self.election_tick = None
+        self.dispatcher_heartbeat_period = None
+        self.node_cert_expiry = None
+        self.external_cas = None
+        self.name = None
+        self.labels = None
+        self.log_driver = None
+        self.signing_ca_cert = None
+        self.signing_ca_key = None
+        self.ca_force_rotate = None
+        self.autolock_managers = None
+        self.rotate_worker_token = None
+        self.rotate_manager_token = None
+
+        for key, value in client.module.params.items():
+            setattr(self, key, value)
+
+        self.update_parameters(client)
+
+    def update_parameters(self, client):
+        self.spec = client.create_swarm_spec(
+            snapshot_interval=self.snapshot_interval,
+            task_history_retention_limit=self.task_history_retention_limit,
+            keep_old_snapshots=self.keep_old_snapshots,
+            log_entries_for_slow_followers=self.log_entries_for_slow_followers,
+            heartbeat_tick=self.heartbeat_tick,
+            election_tick=self.election_tick,
+            dispatcher_heartbeat_period=self.dispatcher_heartbeat_period,
+            node_cert_expiry=self.node_cert_expiry,
+            name=self.name,
+            labels=self.labels,
+            signing_ca_cert=self.signing_ca_cert,
+            signing_ca_key=self.signing_ca_key,
+            ca_force_rotate=self.ca_force_rotate,
+            autolock_managers=self.autolock_managers,
+            log_driver=self.log_driver
+        )
+
+
+class SwarmManager(DockerBaseClass):
+
+    def __init__(self, client, results):
+
+        super(SwarmManager, self).__init__()
+
+        self.client = client
+        self.results = results
+        self.check_mode = self.client.check_mode
+
+        self.parameters = TaskParameters(client)
+
+    def __call__(self):
+        choice_map = {
+            "present": self.init_swarm,
+            "join": self.join,
+            "absent": self.leave,
+            "remove": self.remove,
+            "inspect": self.inspect_swarm
+        }
+
+        choice_map.get(self.parameters.state)()
+
+    def fail(self, msg):
+        self.client.module.fail_json(msg=msg)
```

Comment:
Wrap the contents of `msg` in `to_native` to ensure it's the proper encoding based on the Python version.

Reply:
Good catch, thanks @samdoran

Reply:
@samdoran done !